sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -446,29 +449,46 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
       table
     } else {
       getProviderFromTableProperties(table).map { provider =>
-        assert(provider != "hive", "Hive serde table should not save provider in table properties.")
-        // SPARK-15269: Persisted data source tables always store the location URI as a storage
-        // property named "path" instead of standard Hive `dataLocation`, because Hive only
-        // allows directory paths as location URIs while Spark SQL data source tables also
-        // allows file paths. So the standard Hive `dataLocation` is meaningless for Spark SQL
-        // data source tables.
-        // Spark SQL may also save external data source in Hive compatible format when
-        // possible, so that these tables can be directly accessed by Hive. For these tables,
-        // `dataLocation` is still necessary. Here we also check for input format because only
-        // these Hive compatible tables set this field.
-        val storage = if (table.tableType == EXTERNAL && table.storage.inputFormat.isEmpty) {
-          table.storage.copy(locationUri = None)
+        if (provider == "hive") {
+          val schemaFromTableProps = getSchemaFromTableProperties(table)
+          if (DataType.equalsIgnoreCaseAndNullability(schemaFromTableProps, table.schema)) {
```

Comment:
Schema includes partitioning columns, but it does not include the info of `bucketSpec`. Based on the previous answer about `bucketSpec`, I am not sure whether we also need to check whether `bucketSpec` is the same? If not, should we leave a TODO here; otherwise, we might forget it when we support `bucketSpec` for Hive serde tables.